spec/unit/provider/service/systemd_service_spec.rb

```diff
@@ -297,20 +297,80 @@ def setup_current_resource
         end
       end
 
+      enabled_and_active = {
+        "UnitFileState" => "enabled",
+        "ActiveState" => "active",
+      }
+      disabled_and_inactive = {
+        "UnitFileState" => "disabled",
+        "ActiveState" => "inactive",
+      }
+      # No unit known for this service, and inactive
+      nil_and_inactive = {
+        "ActiveState" => "inactive",
+        "UnitFileState" => nil,
+      }
+      describe "systemd_service_status" do
```

Comment:
The fact that this method exists is an implementation detail, no one outside of this class should care that it exists or be calling it, so explicitly testing it seems like the wrong thing to do here. 

A case could be made for keeping the tests for the error cases though.

Reply:
The intent here is to be testing `systemd_service_status` and how it handles the various outputs and exit codes of `systemctl`; the other tests (e.g. `is_active?`) don't know anything about `systemctl` or its quirks - but that's by design.

Reply:
`systemd_service_status` is not the interface we care about, so it is not important if it changes as long as the methods that depend on it still work, since they are the interfaces we care about. The danger of the tests as they are written in this PR is that changing the implementation of `systemd_service_status` will require updating the tests of `is_active?` and friends, but that won't necessarily be apparent because they will continue to work thanks to the stubs hiding the actual changes.

I agree that all of `is_active?` and similar should not test every detail like the error conditions, those can remain as tests on `systemd_service_status`.

Reply:
I think you're suggesting to do the following, correct?
```
  describe "is_active?" do
  ...
     it "should return true if service is active" do
       systemd_stdout = <<-STDOUT
          ActiveState=active
          UnitFileState=enabled
       STDOUT
       allow(provider).to receive(:shell_out!).with("#{systemctl_path} --system show -p UnitFileState -p ActiveState #{service_name}", {}).and_return(double(stdout: systemd_stdout, exitstatus: 0, error?: false))
       expect(provider.is_active?).to be true
    end
  end
  ```
  
  I can do this, but it'll mean each of the subsequent tests will be explicitly bound to the implementation of `systemd_service_status`; each test will be running it, and a change to the shellout will require a corresponding change to each test (which seems to be at odds with the concern you raised).

Reply:
You're right, these tests are going to be coupled to something because these are unit tests that test running a faked shell command, it's an issue of where the seams are. I don't like coupling the predicate methods so closely to the internal implementation, but I also don't like the scenario you pointed out of having to copy-paste the shell_out stub into each test. In that case, you could make the tests change just the output:

```
  def with_systemd_status(text)
    expect(provider).to receive(:shell_out!).with("#{systemctl_path} --system show -p UnitFileState -p ActiveState #{service_name}", {}).and_return(double(stdout: text, exitstatus: 0, error?: false))
  end

  describe "is_active?" do
  ...
     it "should return true if service is active" do
       with_systemd_status <<~STDOUT
          ActiveState=active
          UnitFileState=enabled
       STDOUT

       expect(provider.is_active?).to be true
    end
  end
```

That could be further encapsulated by reusing the output strings, etc.

Another thought I had is to reorient the tests so that they aren't grouped by method call but by status, so you could do something like:

```
context "when active and enabled" do
  before do
    with_systemd_status <<~STDOUT
      ActiveState=active
      UnitFileState=enabled
    STDOUT
  end
  it "is active" do
   expect(provider.is_active?).to eq(true)
  end

  it "is not masked" do
    expect(provider.is_masked?).to eq(false)
  end
end
```

This could make the tests harder to understand, it was just a stray thought I had.

Reply:
I've fixed the tests to all mock the shellout with a function per your suggestion.